src/metabase/pulse.clj

```diff
@@ -345,117 +376,119 @@
+  {:arglists '([alert-or-pulse parts channel])}
   (fn [pulse _ {:keys [channel_type]}]
     [(alert-or-pulse pulse) (keyword channel_type)]))
 
 (defmethod notification [:pulse :email]
-  [{pulse-id :id, pulse-name :name, dashboard-id :dashboard_id, :as pulse} results {:keys [recipients]}]
+  [{pulse-id :id, pulse-name :name, dashboard-id :dashboard_id, :as pulse} parts {:keys [recipients]}]
   (log/debug (u/format-color 'cyan (trs "Sending Pulse ({0}: {1}) with {2} Cards via email"
-                                        pulse-id (pr-str pulse-name) (count results))))
+                                        pulse-id (pr-str pulse-name) (parts->cards-count parts))))
   (let [email-recipients (filterv u/email? (map :email recipients))
-        query-results    (filter :card results)
-        timezone         (-> query-results first :card defaulted-timezone)
+        card-part        (filter :card parts)
+        timezone         (-> card-part first :card defaulted-timezone)
         dashboard        (update (t2/select-one Dashboard :id dashboard-id) :description markdown/process-markdown :html)]
     {:subject      (subject pulse)
      :recipients   email-recipients
      :message-type :attachments
-     :message      (messages/render-pulse-email timezone pulse dashboard results)}))
+     :message      (messages/render-pulse-email timezone pulse dashboard parts)}))
 
 (defmethod notification [:pulse :slack]
   [{pulse-id :id, pulse-name :name, dashboard-id :dashboard_id, :as pulse}
-   results
+   parts
    {{channel-id :channel} :details}]
   (log/debug (u/format-color 'cyan (trs "Sending Pulse ({0}: {1}) with {2} Cards via Slack"
-                                        pulse-id (pr-str pulse-name) (count results))))
+                                        pulse-id (pr-str pulse-name) (parts->cards-count parts))))
   (let [dashboard (t2/select-one Dashboard :id dashboard-id)]
     {:channel-id  channel-id
      :attachments (remove nil?
                           (flatten [(slack-dashboard-header pulse dashboard)
-                                    (create-slack-attachment-data results)
+                                    (create-slack-attachment-data parts)
                                     (when dashboard (slack-dashboard-footer pulse dashboard))]))}))
 
 (defmethod notification [:alert :email]
-  [{:keys [id] :as pulse} results channel]
+  [{:keys [id] :as pulse} parts channel]
   (log/debug (trs "Sending Alert ({0}: {1}) via email" id name))
   (let [condition-kwd    (messages/pulse->alert-condition-kwd pulse)
         email-subject    (trs "Alert: {0} has {1}"
                               (first-question-name pulse)
                               (alert-condition-type->description condition-kwd))
         email-recipients (filterv u/email? (map :email (:recipients channel)))
-        first-result     (first results)
-        timezone         (-> first-result :card defaulted-timezone)]
+        first-part       (first parts)
```

Comment:
A `part` might not be a card type. Should this be the same as line 418? https://github.com/metabase/metabase/pull/30478/files#diff-a780e860575d9afab3850662e50f677408d861f5759cc91c17e985f023e7fed4R418

Reply:
I think under the context of alert it'll be a card. but yea doing a `(comp first filter) not literally` is tighter, will change.